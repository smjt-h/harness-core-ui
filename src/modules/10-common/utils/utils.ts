/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export const isCommunityPlan = (): boolean => false

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined
