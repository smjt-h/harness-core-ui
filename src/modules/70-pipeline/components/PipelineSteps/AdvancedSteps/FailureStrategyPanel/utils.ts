/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import type { FormikErrors, FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'
import { Intent } from '@blueprintjs/core'

import type {
  RetryFailureActionConfig,
  IgnoreFailureActionConfig,
  ManualInterventionFailureActionConfig,
  AbortFailureActionConfig,
  StepGroupFailureActionConfig,
  MarkAsSuccessFailureActionConfig,
  StageRollbackFailureActionConfig,
  FailureStrategyConfig,
  OnFailureConfig
} from 'services/cd-ng'

export type AllActions =
  | RetryFailureActionConfig
  | IgnoreFailureActionConfig
  | ManualInterventionFailureActionConfig
  | AbortFailureActionConfig
  | StepGroupFailureActionConfig
  | MarkAsSuccessFailureActionConfig
  | StageRollbackFailureActionConfig

export interface AllFailureStrategyConfig extends FailureStrategyConfig {
  onFailure: OnFailureConfig & { action: AllActions }
}

export function hasItems<T>(data?: T[]): boolean {
  return Array.isArray(data) && data.length > 0
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function findTabWithErrors(errors?: FormikErrors<any>): number {
  const fsErrors = errors?.failureStrategies

  return Array.isArray(fsErrors)
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fsErrors.findIndex((err: any) => !isEmpty(err))
    : -1
}

export interface FormState {
  failureStrategies?: AllFailureStrategyConfig[]
}

export interface HandleChangeInStrategiesProps {
  formValues: FormState
  selectedStrategyNum: number
  setSelectedStrategyNum: (value: SetStateAction<number>) => void
  setFormikState: FormikProps<FormState>['setFormikState']
}

export function handleChangeInStrategies({
  formValues,
  selectedStrategyNum,
  setSelectedStrategyNum,
  setFormikState
}: HandleChangeInStrategiesProps): void {
  const fs = formValues.failureStrategies
  /* istanbul ignore else */
  if (Array.isArray(fs)) {
    /* istanbul ignore else */
    if (selectedStrategyNum >= fs.length) {
      // select the new last tab, if the last tab was deleted
      setSelectedStrategyNum(Math.max(0, fs.length - 1))
    }

    /* istanbul ignore else */
    if (fs.length === 0) {
      // reset errors when all the tabs are deleted
      setFormikState({ errors: {}, submitCount: 0 })
    }
  }
}

export function getTabIntent(i: number, selectedStrategyNum: number): Intent {
  return i === selectedStrategyNum ? Intent.PRIMARY : Intent.NONE
}
