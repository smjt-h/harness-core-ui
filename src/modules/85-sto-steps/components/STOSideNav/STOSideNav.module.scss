/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.sideNavTabs {
  font-size: 12px !important;
  :global {
    .bp3-tab-list {
      align-items: center;
      margin: var(--spacing-medium);
      margin-right: 30px;
      .bp3-tab {
        color: white !important;
      }
    }
  }
}

.projectPanel {
  margin-top: -20px !important; // Compensate for excessive project select padding
}
