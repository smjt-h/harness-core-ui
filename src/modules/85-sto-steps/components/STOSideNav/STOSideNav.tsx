/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { Layout } from '@wings-software/uicore'
import { Tabs, Tab } from '@blueprintjs/core'
import routes from '@common/RouteDefinitions'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { SidebarLink } from '@common/navigation/SideNav/SideNav'
import { ProjectSelector } from '@projects-orgs/components/ProjectSelector/ProjectSelector'
import { useAppStore } from 'framework/AppStore/AppStoreContext'
import { useStrings } from 'framework/strings'
import css from './STOSideNav.module.scss'

export default function STOSideNav(): React.ReactElement {
  const { getString } = useStrings()
  const { accountId, projectIdentifier } = useParams<ProjectPathProps>()
  const { selectedProject } = useAppStore()
  const history = useHistory()
  const [isProjectMode, setIsProjectMode] = useState<boolean>(!!projectIdentifier)
  // Telemetry?

  const toProjectMode = () => {
    setIsProjectMode(true)
    if (selectedProject?.identifier) {
      history.push(
        routes.toSTOProjectOverview({
          accountId,
          projectIdentifier: selectedProject.identifier,
          orgIdentifier: selectedProject.orgIdentifier || /* istanbul ignore next */ ''
        })
      )
    }
  }

  const toAccountMode = () => {
    setIsProjectMode(false)
    history.push(routes.toSTOOverview({ accountId }))
  }

  return (
    <Layout.Vertical spacing="small">
      <Tabs id="navTab" selectedTabId={isProjectMode ? 'project' : 'account'} className={css.sideNavTabs}>
        <Tab id="account" title={getString('account')} onClickCapture={toAccountMode} panel={<TabPanel />} />
        <Tabs.Expander />
        <Tab
          id="project"
          title={getString('projectLabel')}
          onClickCapture={toProjectMode}
          panel={<TabPanel isProjectMode />}
        />
      </Tabs>
    </Layout.Vertical>
  )
}

interface TabPanelProps {
  isProjectMode?: boolean
}

const TabPanel: React.FC<TabPanelProps> = ({ isProjectMode }) => {
  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()
  const { getString } = useStrings()
  const { updateAppStore } = useAppStore()
  const history = useHistory()
  // Telemetry?

  const showLinks = !isProjectMode || projectIdentifier

  return (
    <Layout.Vertical spacing="small" className={isProjectMode ? css.projectPanel : undefined}>
      {isProjectMode && (
        <ProjectSelector
          onSelect={selectedProject => {
            updateAppStore({ selectedProject: selectedProject })
            history.push(
              routes.toSTOProjectOverview({
                accountId,
                projectIdentifier: selectedProject.identifier,
                orgIdentifier: selectedProject.orgIdentifier || /* istanbul ignore next */ ''
              })
            )
          }}
        />
      )}

      {showLinks && (
        <React.Fragment>
          {isProjectMode ? (
            <SidebarLink
              label={getString('overview')}
              to={routes.toSTOProjectOverview({
                accountId,
                projectIdentifier,
                orgIdentifier
              })}
            />
          ) : (
            <SidebarLink label={getString('overview')} to={routes.toSTOOverview({ accountId })} />
          )}
        </React.Fragment>
      )}
    </Layout.Vertical>
  )
}
