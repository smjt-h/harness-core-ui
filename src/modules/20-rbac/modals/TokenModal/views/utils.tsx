/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getReadableDateTime } from '@common/utils/dateUtils'

export const DATE_FORMAT = 'MM/DD/YYYY'

export const hasNoExpiryDate = (value?: number): boolean => {
  return getReadableDateTime(value, DATE_FORMAT) === '12/31/2099'
}

export const getSelectedExpiration = (value?: number): string => {
  if (!value) {
    return '30'
  } else if (hasNoExpiryDate(value)) {
    return '-1'
  }
  return 'custom'
}

export const getSelectedExpirationDate = (value: string): string => {
  switch (value) {
    case '30':
    case '90':
    case '180':
      return getReadableDateTime(new Date().setDate(new Date().getDate() + parseInt(value)), DATE_FORMAT)
    case '-1':
      return '12/31/2099'
    default:
      return ''
  }
}

export const getExpirationDate = (value?: number): string => {
  if (value) {
    return getReadableDateTime(value, DATE_FORMAT)
  } else {
    return getSelectedExpirationDate(getSelectedExpiration(value))
  }
}
