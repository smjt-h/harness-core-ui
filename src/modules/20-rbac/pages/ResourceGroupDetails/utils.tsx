/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import produce from 'immer'
import type { SelectOption } from '@wings-software/uicore'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import { isDynamicResourceSelector, SelectionType } from '@rbac/utils/utils'
import type {
  ResourceSelector,
  ResponseResourceTypeDTO,
  ScopeSelector,
  ResourceGroupV2,
  ResourceSelectorV2,
  ResourceGroupV2Request,
  ResourceFilter
} from 'services/resourcegroups'
import { Scope } from '@common/interfaces/SecretsInterface'
import type { UseStringsReturn } from 'framework/strings'
import RbacFactory from '@rbac/factories/RbacFactory'
import { getScopeFromDTO } from '@common/components/EntityReference/EntityReference'

export enum SelectorScope {
  CURRENT = 'CURRENT',
  INCLUDE_CHILD_SCOPES = 'INCLUDE_CHILD_SCOPES',
  CUSTOM = 'CUSTOM'
}

enum ValidatorTypes {
  BY_RESOURCE_IDENTIFIER = 'BY_RESOURCE_IDENTIFIER',
  BY_RESOURCE_TYPE = 'BY_RESOURCE_TYPE',
  BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES = 'BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES'
}

export const getSelectedResourcesMap = (
  resourceTypes: ResourceType[],
  resourceFilter?: ResourceFilter
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()
  if (resourceFilter?.includeAllResources) {
    resourceTypes.forEach(resourceType => {
      map.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  } else {
    resourceFilter?.resources?.map(resourceSelector => {
      if (resourceSelector.identifiers?.length) {
        map.set(resourceSelector.resourceType as ResourceType, resourceSelector.identifiers)
      } else {
        map.set(resourceSelector.resourceType as ResourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      }
    })
  }

  return map
}

export const validateResourceSelectors = (value: ResourceSelector[]): ResourceType[] => {
  return value
    .filter(val => {
      if (!isDynamicResourceSelector(val.type)) {
        if (val.identifiers?.length < 1) {
          return true
        }
      }
      return false
    })
    .map(val => val.resourceType)
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelectorV2[] => {
  const resourceSelectors: ResourceSelectorV2[] = []
  map.forEach((value, key) => {
    if (isDynamicResourceSelector(value)) {
      resourceSelectors.push({
        resourceType: key
      })
    } else {
      resourceSelectors.push({
        resourceType: key,
        identifiers: value as string[]
      })
    }
  })
  return resourceSelectors
}

export const computeResourceMapOnChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  resourceType: ResourceType,
  isAdd: boolean,
  identifiers?: string[]
): void => {
  if (identifiers) {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, identifiers)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          if (Array.isArray(draft.get(resourceType))) {
            draft.set(
              resourceType,
              (draft.get(resourceType) as string[]).filter(el => !identifiers.includes(el))
            )
          }
          if (draft.get(resourceType)?.length === 0) {
            draft.delete(resourceType)
          }
        })
      )
    }
  } else {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.delete(resourceType)
        })
      )
    }
  }
}

export const computeResourceMapOnMultiChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  types: ResourceType[],
  isAdd: boolean
): void => {
  if (isAdd) {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      })
    )
  } else {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.delete(resourceType)
        })
      })
    )
  }
}

export const getScopeLabelFromApi = (
  getString: UseStringsReturn['getString'],
  scope: Scope,
  resourceGroup: ResourceGroupV2
): string => {
  const selectorScope = getScopeType(resourceGroup)
  const dropDownItems = getScopeDropDownItems(scope, getString)
  const option = dropDownItems.filter(item => item.value === selectorScope)
  return option.length ? option[0].label : ''
}

export const getScopeDropDownItems = (scope: Scope, getString: UseStringsReturn['getString']): SelectOption[] => {
  switch (scope) {
    case Scope.PROJECT:
      return [{ label: getString('rbac.scopeItems.projectOnly'), value: SelectorScope.CURRENT }]
    case Scope.ORG:
      return [
        { label: getString('rbac.scopeItems.orgOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.orgAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
    case Scope.ACCOUNT:
    default:
      return [
        { label: getString('rbac.scopeItems.accountOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.accountAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
  }
}

export const getSelectionType = (resourceGroup?: ResourceGroupV2): SelectionType => {
  if (resourceGroup?.resourceFilter?.includeAllResources) {
    return SelectionType.ALL
  }
  return SelectionType.SPECIFIED
}
export const getScopeType = (resourceGroup?: ResourceGroupV2): SelectorScope => {
  if (resourceGroup?.includedScopes?.length) {
    for (const scope of resourceGroup?.includedScopes) {
      if (getScopeFromDTO(resourceGroup) === getScopeFromDTO(scope)) {
        return scope.filter === 'INCLUDING_CHILD_SCOPES' ? SelectorScope.INCLUDE_CHILD_SCOPES : SelectorScope.CURRENT
      } else {
        return SelectorScope.CUSTOM
      }
    }
  }

  return SelectorScope.CURRENT
}

export const getFormattedDataForApi = (
  data: ResourceGroupV2,
  selectionType: SelectionType,
  includedScopes: ScopeSelector[],
  resourceSelectors: ResourceSelectorV2[]
): ResourceGroupV2Request => {
  return {
    resourceGroup: {
      ...data,
      resourceFilter:
        selectionType === SelectionType.ALL
          ? {
              includeAllResources: true
            }
          : {
              resources: resourceSelectors,
              includeAllResources: false
            },
      includedScopes: includedScopes === [] ? getIncludedScopes(SelectorScope.CURRENT, data) : includedScopes
    }
  }
}

export const getFilteredResourceTypes = (
  data: ResponseResourceTypeDTO | null,
  selectedScope: SelectorScope
): ResourceType[] => {
  const resourceTypes = data?.data?.resourceTypes || []
  return resourceTypes?.reduce((acc: ResourceType[], value) => {
    if (
      value.name &&
      RbacFactory.isRegisteredResourceType(value.name as ResourceType) &&
      ((selectedScope === SelectorScope.CURRENT && value.validatorTypes?.includes(ValidatorTypes.BY_RESOURCE_TYPE)) ||
        selectedScope !== SelectorScope.CURRENT)
    ) {
      acc.push(value.name as ResourceType)
    }
    return acc
  }, [])
}

export const cleanUpResourcesMap = (
  resourceTypes: ResourceType[],
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  selectionType: SelectionType
): Map<ResourceType, string | string[]> => {
  const types = [...resourceTypes]
  selectedResourcesMap.forEach((_value, key) => {
    if (!types.includes(key)) {
      selectedResourcesMap.delete(key)
    } else {
      types.splice(types.indexOf(key), 1)
      selectedResourcesMap.set(key, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    }
  })
  if (selectionType === SelectionType.ALL) {
    types.map(resourceType => {
      selectedResourcesMap.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  }
  return selectedResourcesMap
}

export const getDefaultIncludedScope = (
  accountIdentifier: string,
  orgIdentifier?: string,
  projectIdentifier?: string,
  includedScopes?: ScopeSelector[]
): ScopeSelector[] => {
  if (includedScopes?.length) {
    return includedScopes
  }
  return [
    {
      accountIdentifier,
      orgIdentifier,
      projectIdentifier,
      filter: 'EXCLUDING_CHILD_SCOPES'
    }
  ]
}

export const getIncludedScopes = (type: SelectorScope, resourceGroup: ResourceGroupV2): ScopeSelector[] => {
  const { accountIdentifier, orgIdentifier, projectIdentifier } = resourceGroup
  switch (type) {
    case SelectorScope.CUSTOM:
      //Replace with custom scope once it is added
      return []
    case SelectorScope.INCLUDE_CHILD_SCOPES:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'INCLUDING_CHILD_SCOPES'
        }
      ]
    default:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'EXCLUDING_CHILD_SCOPES'
        }
      ]
  }
}
