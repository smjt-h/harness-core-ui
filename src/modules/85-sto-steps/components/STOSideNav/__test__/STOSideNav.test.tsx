/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, waitFor, fireEvent } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import STOSideNav from '../STOSideNav'

const testPathOverview = routes.toSTOOverview({ accountId: ':accountId' })
const testPathProjectOverview = routes.toSTOProjectOverview({
  accountId: ':accountId',
  orgIdentifier: ':orgIdentifier',
  projectIdentifier: ':projectIdentifier'
})
//'/account/:accountId/sto/orgs/:orgIdentifier/projects/:projectIdentifier/overview'
const testParams = {
  accountId: 'accountId',
  orgIdentifier: 'orgIdentifier',
  projectIdentifier: 'projectIdentifier',
  module: 'sto'
}

jest.mock('@projects-orgs/components/ProjectSelector/ProjectSelector', () => ({
  ProjectSelector: function ProjectSelectorComp(props: any) {
    return (
      <button
        onClick={() => props.onSelect({ identifier: 'newProject', orgIdentifier: 'newOrg' })}
        id="projectSelector"
      />
    )
  }
}))

describe('STO side nav tests', () => {
  test('renders account path', () => {
    const { container } = render(
      <TestWrapper path={testPathOverview} pathParams={testParams}>
        <STOSideNav />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('renders project path', () => {
    const { container } = render(
      <TestWrapper path={testPathProjectOverview} pathParams={testParams}>
        <STOSideNav />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('switches to project mode when no project in app store', async () => {
    const { container } = render(
      <TestWrapper path={testPathOverview} pathParams={testParams}>
        <STOSideNav />
      </TestWrapper>
    )

    const projectTab = await waitFor(() => container.querySelector('[data-tab-id="project"]'))
    fireEvent.click(projectTab!)

    expect(projectTab).toHaveAttribute('aria-selected', 'true')
  })

  test('switches to project overview when app store has project', async () => {
    const { container, getByTestId } = render(
      <TestWrapper
        path={testPathOverview}
        pathParams={testParams}
        defaultAppStoreValues={{ selectedProject: { name: 'Target', identifier: 'PROJECT', orgIdentifier: 'ORG' } }}
      >
        <STOSideNav />
      </TestWrapper>
    )

    const projectTab = await waitFor(() => container.querySelector('[data-tab-id="project"]'))
    fireEvent.click(projectTab!)

    expect(getByTestId('location').innerHTML).toEqual(
      routes.toSTOProjectOverview({
        accountId: 'accountId',
        orgIdentifier: 'ORG',
        projectIdentifier: 'PROJECT'
      })
    )
  })

  test('switches to project overview when project selected', async () => {
    const { container, getByTestId } = render(
      <TestWrapper path={testPathOverview} pathParams={testParams}>
        <STOSideNav />
      </TestWrapper>
    )

    const projectSelector = await waitFor(() => container.querySelector('#projectSelector'))
    fireEvent.click(projectSelector!)

    expect(getByTestId('location').innerHTML).toEqual(
      routes.toSTOProjectOverview({
        accountId: 'accountId',
        orgIdentifier: 'newOrg',
        projectIdentifier: 'newProject'
      })
    )
  })

  test('switches to account overview', async () => {
    const { container, getByTestId } = render(
      <TestWrapper
        path={testPathProjectOverview}
        pathParams={testParams}
        defaultAppStoreValues={{ selectedProject: { name: 'Target', identifier: 'PROJECT', orgIdentifier: 'ORG' } }}
      >
        <STOSideNav />
      </TestWrapper>
    )

    const projectTab = await waitFor(() => container.querySelector('[data-tab-id="account"]'))
    fireEvent.click(projectTab!)

    expect(getByTestId('location').innerHTML).toEqual(routes.toSTOOverview({ accountId: 'accountId' }))
  })
})
