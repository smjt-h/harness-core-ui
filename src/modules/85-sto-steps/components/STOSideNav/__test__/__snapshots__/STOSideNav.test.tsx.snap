// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STO side nav tests renders account path 1`] = `
<div>
  <div
    class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
  >
    <div
      class="bp3-tabs sideNavTabs"
    >
      <div
        class="bp3-tab-list"
        role="tablist"
      >
        <div
          class="bp3-tab-indicator-wrapper"
          style="height: 0px; transform: translateX(0px) translateY(0px); width: 0px;"
        >
          <div
            class="bp3-tab-indicator"
          />
        </div>
        <div
          aria-controls="bp3-tab-panel_navTab_account"
          aria-disabled="false"
          aria-expanded="true"
          aria-selected="true"
          class="bp3-tab"
          data-tab-id="account"
          id="bp3-tab-title_navTab_account"
          role="tab"
          tabindex="0"
        >
          account
        </div>
        <div
          class="bp3-flex-expander"
        />
        <div
          aria-controls="bp3-tab-panel_navTab_project"
          aria-disabled="false"
          aria-expanded="false"
          aria-selected="false"
          class="bp3-tab"
          data-tab-id="project"
          id="bp3-tab-title_navTab_project"
          role="tab"
          tabindex="0"
        >
          projectLabel
        </div>
      </div>
      <div
        aria-hidden="false"
        aria-labelledby="bp3-tab-title_navTab_account"
        class="bp3-tab-panel"
        id="bp3-tab-panel_navTab_account"
        role="tabpanel"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
        >
          <a
            aria-current="page"
            class="link selected"
            href="/account/accountId/sto/overview"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              overview
            </p>
          </a>
          <a
            class="link"
            href="/account/accountId/sto/targets"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              stoSteps.targets.testTargets
            </p>
          </a>
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="bp3-tab-title_navTab_project"
        class="bp3-tab-panel"
        id="bp3-tab-panel_navTab_project"
        role="tabpanel"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-small StyledProps--main projectPanel"
        >
          <button
            id="projectSelector"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`STO side nav tests renders project path 1`] = `
<div>
  <div
    class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
  >
    <div
      class="bp3-tabs sideNavTabs"
    >
      <div
        class="bp3-tab-list"
        role="tablist"
      >
        <div
          class="bp3-tab-indicator-wrapper"
          style="height: 0px; transform: translateX(0px) translateY(0px); width: 0px;"
        >
          <div
            class="bp3-tab-indicator"
          />
        </div>
        <div
          aria-controls="bp3-tab-panel_navTab_account"
          aria-disabled="false"
          aria-expanded="false"
          aria-selected="false"
          class="bp3-tab"
          data-tab-id="account"
          id="bp3-tab-title_navTab_account"
          role="tab"
          tabindex="0"
        >
          account
        </div>
        <div
          class="bp3-flex-expander"
        />
        <div
          aria-controls="bp3-tab-panel_navTab_project"
          aria-disabled="false"
          aria-expanded="true"
          aria-selected="true"
          class="bp3-tab"
          data-tab-id="project"
          id="bp3-tab-title_navTab_project"
          role="tab"
          tabindex="0"
        >
          projectLabel
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="bp3-tab-title_navTab_account"
        class="bp3-tab-panel"
        id="bp3-tab-panel_navTab_account"
        role="tabpanel"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
        >
          <a
            class="link"
            href="/account/accountId/sto/overview"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              overview
            </p>
          </a>
          <a
            class="link"
            href="/account/accountId/sto/targets"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              stoSteps.targets.testTargets
            </p>
          </a>
        </div>
      </div>
      <div
        aria-hidden="false"
        aria-labelledby="bp3-tab-title_navTab_project"
        class="bp3-tab-panel"
        id="bp3-tab-panel_navTab_project"
        role="tabpanel"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-small StyledProps--main projectPanel"
        >
          <button
            id="projectSelector"
          />
          <a
            aria-current="page"
            class="link selected"
            href="/account/accountId/sto/orgs/orgIdentifier/projects/projectIdentifier/overview"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              overview
            </p>
          </a>
          <a
            class="link"
            href="/account/accountId/sto/orgs/orgIdentifier/projects/projectIdentifier/targets"
          >
            <p
              class="StyledProps--font StyledProps--main text"
            >
              stoSteps.targets.testTargets
            </p>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
